### ISSUE TITLE
Publication error event not sent when RabbitMQ is not up

### ISSUE THREAD

**Issue by AnthonySendra** [guidesmiths/rascal#88]
Hello,

I'm having trouble with RabbitMQ and error management. Here's how to reproduce:

1. Run container RabbitMQ
2. Run my application with broker connect
3. Shutdown RabbitMQ
4. Trigger the function that send a message to RabbitMQ

Here is my code that is triggered:

```js
async myFunc() {
  try {
    const publication = await this.broker.publish("toto", msg);
    publication
      .on("error", (err, messageId) => {
        console.error("Publisher error", err, messageId);
      })
      .on("success", messagesId => {
        console.log("success", messagesId);
      });
  } catch (error) {
      console.error("Catch error", error);
  }
}
```
Unfortunately, my `console.error("Publisher error", err, messageId);` is never executed.
But somewhere else I have `broker.on("error", console.error);` just after the `connect`, and my error is logged here.

My problem is I'd like to know there's an error when I'm executing my "myFunc" function. This function is executed in a API and I'd like to trace the error back to the user when RabbitMQ is inaccessible.
Or even better, maybe there is a way to find the connection status for a vhost?

Thanks!

**Comment by cressie176** [comment]
Hi @AnthonySendra,

I don't think what you want is currently possible. When a connection error occurs, rascal pauses the underlying channel pools, causing requests for new channels to be queued. This means that the publish function will not return, until the connection has been re-established. There's not currently a way you can determine the status of the publication or associated vhost. 

I think this would be a good feature to add, but wouldn't be completely reliable - it would be possible for the connection error to occur after checking the status, but before acquiring the channel. I'll give the problem some thought and see what I can come up with.

**Comment by cressie176** [comment]
Hi @AnthonySendra,

FYI I haven't forgotten about this. The reason for the delay, is that I recently added a publication timeout attribute, which can be used to guard against the broker failing to acknowledge messages sent using a confirm channel. This is important since unlike a connection issue, it can be permanent, and without the timeout rascal will never yield, forever blocking that flow of execution.

A general publication timeout would still useful, but only in certain circumstances, such as your API example. It's potentially undesirable for background jobs, which may not be time dependent. Adding two different timeouts is rather unwieldy, so I'm still mulling over what to do.

One option I quite like is to emit a `paused` event from the publication whenever the vhost pauses channel allocation (which is does when shutting down or on connection errors). I could also provide a function to drop the message rather than publishing it when channel allocation was resumed. This would at least allow an API client to return a 50X response, and be confident the message hadn't been published.

I would immediately emit this event on registering the 'paused' event listener if channel allocation was already paused.

```
function publish(publicationName, content) {
  return new Promise(resolve, reject) {
    const publication = await broker.publish(publicationName, content);
    publication.on('success', (messageId) => {
      resolve();
    });
    publication.on('return', (message) => {
      reject(new Error(`Message ${message.properties.messageId} was returned`));
    });
    publication.on('error', (err, messageId) => {
      reject(err);
    });
    publication.on('paused', (messageId) => {
      publication.abort();
      reject(new Error(`Message ${messageId} was aborted`));
    });
  })
};
```

What do you think?

**Comment by AnthonySendra** [comment]
Hey @cressie176, thanks for your answer.

For your implementation problem I don't really know what Rascal's architecture is like, but is the implementation of a health check possible? An event would be sent in case of an error on the health check, which would allow us to implement a circuit breaker.

But your example would also seem to meet my needs. It just doesn't allow me to deal with disconnections at a higher level. But it wouldn't be a big deal.
Just one remark: I don't find the "pause" event very clear. I have the impression that if I don't listen to this event, my message will end up being sent anyway, because it's just "paused". Is this the case?
In any case, if I am able to return a 50X to my client in case of a failure, I would be very happy :)

**Comment by cressie176** [comment]
> is the implementation of a health check possible

Rascal architecture depends heavily on events. The broker already emits events for

- disconnect
- connect
- busy
- ready

disconnect and connect don't currently contain any context, so you can't tell which connection / vhost they relate to. It would be trivial to add the vhost name, but you would still have to work out which publications / subscriptions were affected.

The busy and ready events are emitted when publishing saturates the channel. The events include the vhost name and statistics about the relevant channel pool.

> I don't find the "pause" event very clear. I have the impression that if I don't listen to this event, my message will end up being sent anyway, because it's just "paused". Is this the case?

When publishing a message, Rascal internally queues a request to borrow a channel from a channel pool. If the connection drops, then the queue is paused, so that any pending channel requests will be blocked until the connection has been re-established and the queue resumed. So yes - unless I provide a mechanism to either (a) remove the request from the queue, or (b) discard the request when it is dequeued, the message will end up being sent anyway. This was the idea behind `publication.abort()`.

**Comment by AnthonySendra** [comment]
OK, get it.
Your example looks great to me. If I have both pause event and abort function, I'll be able to do what I want!

**Comment by jtravan3** [comment]
@cressie176 Would it be possible to add an example for the `disconnect` and `connect` events? I think that's exactly what I was looking for but I couldn't find it in the documentation.

**Comment by cressie176** [comment]
Hi @jtravan3,

The `connect` and `disconnect` events were added via a [PR](https://github.com/guidesmiths/rascal/pull/9) years ago and never documented (my bad). I've never actually used them, and on reflection, they're slightly misleading.

* [connect](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Vhost.js#L48) is emitted from the vhost and (and forwarded to the broker) when the vhost is successfully initialised, which involves more than just establishing a connection
* [disconnect](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Vhost.js#L362) is emitted from the vhost (and forwarded to the broker) when the vhost encounters a connection error. Assuming Rascal is configured to retry the connection, you may get multiple `disconnect` events, even though the connection was not successfully established. The disconnect event is not emitted then rascal is shutdown.

Given the above I'll think about whether they should be renamed (maybe `vhost_initialised` and `vhost_connection_error`). You're quite safe to use them though - I won't remove them, but may decide to add new ones that better describe what they do.

**Comment by jtravan3** [comment]
Thanks @cressie176 and after hooking into those events I'm seeing that behavior. For me, that does exactly what I needed it to do, but I'd be happy to see any events you can provide in the future. The more information that I can log, the better.

**Comment by cressie176** [comment]
@jtravan3 I'm looking at the connect/disconnect events now. It looks like `connect` is emitted before events are forwarded from the vhost to the broker, so the very first `connect` event will be lost. I'm planning on introducing a `vhost_initialised` event instead.

```
{
  vhost: 'myVhost',
  connectionUrl: 'amqp//guest:***@localhost:5672?foo=bar'
}
```

For disconnect, the best thing to do is just listen for broker `error` events. I'll add the above payload to those also, i.e.

```
broker.on('error', (err, { vhost, connectionUrl }) => {
  // do stuff
})
```

**Comment by cressie176** [comment]
Unfortunately not only is the connect event emitted before events are forwarded to the broker, it's actually emitted before the user has an opportunity to register a listener. 

The best way I can think of to work around this, is to add an option of not automatically initialising the broker on create. e.g.

```js
Rascal.Broker.create(Object.assign(config, { autostart: false }), function(err, broker) {
  if (err) throw err

  broker.on('vhost_initialised', ({ vhost, connectionUrl }) => {
  })

  broker.init((err) => {    
  })
```
It's pretty ugly and leads to the possibility of people forgetting to initialise the broker, or doing so repeatedly. I'm not sure it's worth it.

**Comment by jtravan3** [comment]
@cressie176 Yeah I noticed that yesterday after I started playing with it more. I was subscribed to both the `error` and `disconnect` events and were getting duplicates when a connection was lost. I also treated the `connect` event like `reconnect` instead since I only got the event after the broker had been initialized.

I, personally, wouldn't use the `autostart: false` so I wouldn't have to manually start the broker but that's just me.

<img width="1245" alt="broker" src="https://user-images.githubusercontent.com/10760045/80914006-b42f0a00-8d16-11ea-924e-2c2082d33e4d.png">

**Comment by cressie176** [comment]
### Publications
I've added a `paused` event and abort method as described above

### Broker Events
I've added `vhost_initialised` (although you still can't get this during startup - it's for reconnections only)

I've also added the vhost details to broker.error events
```js
broker.on('error', (err, { vhost, connectionUrl }) => {
  // ...
});
```

Published as rascal@9.4.0
